File: slidingWindow/slidingWindowMaximum.py
import collections

class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        output  = []
        q = collections.deque() # index
        l = r = 0
        while r < len(nums):
            # pop smaller values from the q
            while q and nums[q[-1]] < nums[r]:
                q.pop()
            q.append(r)
            if l > q[0]:
                q.popleft()
            if (r + 1) >= k:
                output.append(nums[q[0]])
                l += 1
            r += 1
        return output

        
if __name__ == "__main__":
    nums = [1,2,1,0,4,2,6]
    k = 3
    print(Solution().maxSlidingWindow(nums, k))

File: slidingWindow/cppRealizations/slidingWindowMaximum.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;


class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> result;
        deque<int> q;
        int r = 0;
        int l = 0;
        while(r < nums.size()){
            // pop smaller values from the q
            while(!q.empty() && nums[q.back()] < nums[r]){
                q.pop_back();
            }
            q.push_back(r);
            if(l > q.front()){
                q.pop_front();
            }
            if(r + 1 >= k){
                result.push_back(nums[q.front()]);
                l += 1;
            }
            r += 1;
        }
        return result;
    }
};


int main(){
    vector<int> nums = {1,2,1,0,4,2,6};
    int k = 3;
    vector<int> result = Solution().maxSlidingWindow(nums, k);
    for(int i = 0; i < result.size(); i++){
        cout << result[i] << endl;
    }
    return 0;
}